packages/paymaster/src/BiconomyPaymaster.ts

```diff
@@ -1,49 +1,241 @@
+      name: 'BICONOMY',
+      version: '1.0.0'
+    }
+    let webhookData = null
+
+    if (paymasterServiceData.mode) {
+      Logger.log('Requested mode is ', paymasterServiceData.mode)
+      mode = paymasterServiceData.mode
+      // Validation on the mode passed / define allowed enums
+    }
+
+    if (paymasterServiceData.tokenInfo) {
+      if (paymasterServiceData.tokenInfo.preferredToken) {
+        Logger.log('passed preferred token is ', paymasterServiceData.tokenInfo.preferredToken)
+        preferredToken = paymasterServiceData.tokenInfo.preferredToken
+      }
+    }
+
+    Logger.log('userop is ', userOp)
+    // userOp = hexifyUserOp(userOp)
+
+    if (
+      paymasterServiceData.tokenInfo &&
+      paymasterServiceData.tokenInfo.tokenList &&
+      paymasterServiceData.tokenInfo.tokenList.length != 0
+    ) {
+      feeTokensArray = paymasterServiceData.tokenInfo.tokenList
+    }
+
+    const feeQuotes: Array<PaymasterFeeQuote> = []
+
+    if (paymasterServiceData.sponsorshipInfo) {
+      if (paymasterServiceData.sponsorshipInfo.webhookData) {
+        webhookData = paymasterServiceData.sponsorshipInfo.webhookData
+      }
+
+      // Could check here that this must be provided
+      if (paymasterServiceData.sponsorshipInfo.smartAccountInfo) {
+        smartAccountInfo = paymasterServiceData.sponsorshipInfo.smartAccountInfo
+      }
+    }
+
+    try {
+      const response: PaymasterServiceSuccessResponse = await sendRequest({
+        url: `${this.paymasterConfig.paymasterUrl}`,
+        method: HttpMethod.Post,
+        body: {
+          method: 'pm_getFeeQuoteOrData',
+          params: [
+            userOp,
+            {
+              ...(mode !== null && { mode }), // Review can be dynamic
+              tokenInfo: {
+                tokenList: feeTokensArray,
+                ...(preferredToken !== null && { preferredToken })
+              },
+              sponsorshipInfo: {
+                ...(webhookData !== null && { webhookData }),
+                smartAccountInfo: smartAccountInfo
+              }
+            }
+          ], // As per current API
+          id: 4337,
+          jsonrpc: '2.0'
+        }
+      })
+
+      if (response && response.result) {
+        Logger.log('feeInfo ', response.result)
+        if (response.result.mode == 'ERC20') {
+          const feeQuotesResponse: Array<PaymasterFeeQuote> = response.result.feeQuotes
+          const paymasterAddress: string = response.result.paymasterAddress
+          // check all objects iterate and populate below calculation for all tokens
+          return { feeQuotes: feeQuotesResponse, tokenPaymasterAddress: paymasterAddress }
+        } else if (response.result.mode == 'SPONSORED') {
+          const paymasterAndData: string = response.result.paymasterAndData
+          return paymasterAndData
+        } else {
+          // Review
+          return { feeQuotes, tokenPaymasterAddress: '' }
+        }
+      } else {
+        // REVIEW
```

Comment:
throw exception here saying `Failed to fetch feeQuote information`

Reply:
working on it

Reply:
done